src/System.Management.Automation/engine/Modules/ModuleUtils.cs

```diff
@@ -146,59 +168,75 @@ internal static IEnumerable<string> GetDefaultAvailableModuleFiles(bool force, b
-            // if the user give the module path including version, we should be able to find the module as well
-            if (Version.TryParse(fileName, out moduleVersion) && Directory.Exists(Directory.GetParent(directory).ToString()))
-            {
-                fileName = Directory.GetParent(directory).Name;
-            }
-            foreach (var version in GetModuleVersionSubfolders(directory))
+
+            // If the given directory doesn't exist or it's the root folder, then return an empty list.
+            if (!Directory.Exists(directory) || fileName == string.Empty) { return result; }
+
+            // If the user give the module path including version, the module name could be the parent folder name.
+            if (Version.TryParse(fileName, out Version ver))
             {
-                var qualifiedPathWithVersion = Path.Combine(directory, Path.Combine(version.ToString(), fileName));
-                string manifestPath = qualifiedPathWithVersion + StringLiterals.PowerShellDataFileExtension;
-                if (File.Exists(manifestPath))
-                {
-                    bool isValidModuleVersion = version.Equals(ModuleIntrinsics.GetManifestModuleVersion(manifestPath));
+                string parentDirPath = Path.GetDirectoryName(directory);
+                string parentDirName = Path.GetFileName(parentDirPath);
 
-                    if (isValidModuleVersion)
+                // If the parent directory is NOT a root folder, then it could be the module folder.
+                if (parentDirName != string.Empty)
+                {
+                    string manifestPath = Path.Combine(directory, parentDirName);
+                    manifestPath += StringLiterals.PowerShellDataFileExtension;
+                    if (File.Exists(manifestPath) && ver.Equals(ModuleIntrinsics.GetManifestModuleVersion(manifestPath)))
                     {
                         result.Add(manifestPath);
+                        return result;
                     }
                 }
             }
 
-            foreach (string ext in ModuleIntrinsics.PSModuleExtensions)
+            // If we reach here, then use the given directory as the module folder.
+            foreach (var version in GetModuleVersionSubfolders(directory))
             {
-                string moduleFile = Path.Combine(directory, fileName) + ext;
-
-                if (!Utils.FileExists(moduleFile))
+                string manifestPath = Path.Combine(directory, version.ToString(), fileName);
+                manifestPath += StringLiterals.PowerShellDataFileExtension;
+                if (File.Exists(manifestPath) && version.Equals(ModuleIntrinsics.GetManifestModuleVersion(manifestPath)))
                 {
-                    continue;
+                    result.Add(manifestPath);
                 }
+            }
 
-                result.Add(moduleFile);
+            foreach (string ext in ModuleIntrinsics.PSModuleExtensions)
```

Comment:
This logic looks very similar to the `Import-Module` local load logic:
https://github.com/PowerShell/PowerShell/blob/bbb4f2ea84f6c4a8e8775978178826b61db20bda/src/System.Management.Automation/engine/Modules/ImportModuleCommand.cs#L530-L760

Where we first check the path, then the folder, then the versions, then the extensions, etc.